File: code/HuaWeiNote/app/src/main/java/com/fan/hwnote/app/model/NoteRepository.kt
package com.fan.hwnote.app.model

import android.content.ContentValues
import android.content.Context
import android.database.Cursor
import com.fan.hwnote.app.model.db.NoteDbHelper
import com.fan.hwnote.app.model.entity.Block
import com.fan.hwnote.app.model.entity.Note
import com.fan.hwnote.app.model.json.NoteJson
import com.fan.hwnote.app.model.storage.NoteFileStorage
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

/**
 * 数据层统一门面。所有 IO 都是 suspend，内部已切到 Dispatchers.IO。
 *
 * 用法：在 App.onCreate 调一次 NoteRepository.init(this)；之后从任何地方 NoteRepository.list() 等。
 *
 * 单例形态：object，状态用 lateinit 注入 — init 可重复调（覆盖之前的 dbHelper / fileStorage）。
 */
object NoteRepository {

    private lateinit var dbHelper: NoteDbHelper
    private lateinit var fileStorage: NoteFileStorage

    fun init(context: Context) {
        val app = context.applicationContext
        dbHelper = NoteDbHelper(app)
        fileStorage = NoteFileStorage(app)
    }

    suspend fun list(
        filter: ListFilter = ListFilter.All,
        sortBy: SortBy = SortBy.UPDATED_DESC,
        query: String? = null,
    ): List<Note> = withContext(Dispatchers.IO) {
        val orderBy = when (sortBy) {
            SortBy.UPDATED_DESC -> "updated_at DESC"
            SortBy.CREATED_DESC -> "created_at DESC"
        }
        val where = mutableListOf<String>()
        val args = mutableListOf<String>()
        when (filter) {
            ListFilter.All -> where += "deleted_at = 0"
            ListFilter.Uncategorized -> where += "deleted_at = 0 AND notebook_id IS NULL"
            ListFilter.Favorite -> where += "deleted_at = 0 AND is_favorite = 1"
            ListFilter.Deleted -> where += "deleted_at != 0"
            is ListFilter.Folder -> {
                where += "deleted_at = 0 AND notebook_id IN (SELECT id FROM notebooks WHERE folder_id = ? AND deleted_at = 0)"
                args += filter.folderId.toString()
            }
            is ListFilter.Notebook -> {
                where += "deleted_at = 0 AND notebook_id = ?"
                args += filter.notebookId.toString()
            }
        }
        if (!query.isNullOrEmpty()) {
            val like = "%$query%"
            where += "(title LIKE ? OR plain_text LIKE ?)"
            args += like
            args += like
        }
        val selection = where.joinToString(" AND ")
        val out = mutableListOf<Note>()
        val cursor = dbHelper.readableDatabase.query(
            "notes", null, selection, args.toTypedArray(), null, null, orderBy,
        )
        cursor.use { c -> while (c.moveToNext()) out += cursorToNote(c) }
        out
    }

    suspend fun get(id: Long): Note? = withContext(Dispatchers.IO) {
        val db = dbHelper.readableDatabase
        val cursor = db.query("notes", null, "id = ?", arrayOf(id.toString()), null, null, null)
        cursor.use { c ->
            if (!c.moveToFirst()) null else cursorToNote(c)
        }
    }

    /**
     * 保存笔记。
     * - 若 note.id == 0L：插入新行，使用 note.createdAt（>0 时）或 now；返回新生成的 id。
     * - 若 note.id  > 0：UPDATE 该行，createdAt 不动，updatedAt = now；返回原 id。
     */
    suspend fun save(note: Note): Long = withContext(Dispatchers.IO) {
        val now = System.currentTimeMillis()
        val cv = ContentValues().apply {
            put("title", note.title)
            put("plain_text", note.content.toPlainText())
            put("content_json", NoteJson.toJson(note.content))
            put("is_favorite", if (note.isFavorite) 1 else 0)
            put("updated_at", now)
            if (note.categoryId == null) putNull("category_id") else put("category_id", note.categoryId)
            put("deleted_at", note.deletedAt)
            if (note.notebookId == null) putNull("notebook_id") else put("notebook_id", note.notebookId)
        }
        val db = dbHelper.writableDatabase
        if (note.id == 0L) {
            val createdAt = if (note.createdAt > 0) note.createdAt else now
            cv.put("created_at", createdAt)
            db.insert("notes", null, cv)
        } else {
            db.update("notes", cv, "id = ?", arrayOf(note.id.toString()))
            note.id
        }
    }

    /** 软删除：标记 deleted_at = now。文件不动（restore 后还要用）。 */
    suspend fun softDelete(id: Long) = withContext(Dispatchers.IO) {
        val cv = ContentValues().apply { put("deleted_at", System.currentTimeMillis()) }
        dbHelper.writableDatabase.update("notes", cv, "id = ?", arrayOf(id.toString()))
        Unit
    }

    /** 恢复：deleted_at = 0。 */
    suspend fun restore(id: Long) = withContext(Dispatchers.IO) {
        val cv = ContentValues().apply { put("deleted_at", 0L) }
        dbHelper.writableDatabase.update("notes", cv, "id = ?", arrayOf(id.toString()))
        Unit
    }

    /** 彻底删除：DELETE 行 + 删本地文件目录。 */
    suspend fun deletePermanently(id: Long) = withContext(Dispatchers.IO) {
        dbHelper.writableDatabase.delete("notes", "id = ?", arrayOf(id.toString()))
        fileStorage.deleteNoteDir(id)
    }

    /** 清理超 ttlMs 的软删笔记。返回清理的笔记数。 */
    suspend fun purgeExpired(
        now: Long = System.currentTimeMillis(),
        ttlMs: Long = 30L * 24 * 60 * 60 * 1000,
    ): Int = withContext(Dispatchers.IO) {
        val cutoff = now - ttlMs
        val db = dbHelper.writableDatabase
        val ids = mutableListOf<Long>()
        db.query(
            "notes", arrayOf("id"),
            "deleted_at > 0 AND deleted_at < ?", arrayOf(cutoff.toString()),
            null, null, null,
        ).use { c -> while (c.moveToNext()) ids += c.getLong(0) }
        for (id in ids) {
            db.delete("notes", "id = ?", arrayOf(id.toString()))
            fileStorage.deleteNoteDir(id)
        }
        ids.size
    }

    /**
     * 扫描某笔记本地 images/ + audio/ 目录，删除当前 content.blocks 未引用的孤儿文件。
     *
     * 用于 M11 撤销 / 重做：编辑器删块时不立即 rm 文件（保证 undo 能复活），
     * 保存成功后异步清理本次操作产生的真正孤儿。
     *
     * - 容错：任何 IOException 仅 log，不抛
     * - 幂等：多次调用对同一份 note 行为一致
     */
    suspend fun cleanOrphanFiles(noteId: Long, note: Note) =
        withContext(Dispatchers.IO) {
            if (noteId <= 0L) return@withContext
            val referenced = mutableSetOf<String>()
            for (b in note.content.blocks) {
                when (b) {
                    is Block.ImageBlock -> referenced += b.fileName
                    is Block.AudioBlock -> referenced += b.fileName
                    is Block.TextBlock, is Block.ChecklistBlock -> Unit
                }
            }
            runCatching {
                fileStorage.imageDir(noteId).listFiles()?.forEach { f ->
                    if (f.isFile && f.name !in referenced) f.delete()
                }
            }.onFailure { android.util.Log.w("NoteRepository", "cleanOrphan image failed", it) }
            runCatching {
                fileStorage.audioDir(noteId).listFiles()?.forEach { f ->
                    if (f.isFile && f.name !in referenced) f.delete()
                }
            }.onFailure { android.util.Log.w("NoteRepository", "cleanOrphan audio failed", it) }
        }

    suspend fun count(filter: ListFilter): Int = withContext(Dispatchers.IO) {
        val where = mutableListOf<String>()
        val args = mutableListOf<String>()
        when (filter) {
            ListFilter.All -> where += "deleted_at = 0"
            ListFilter.Uncategorized -> where += "deleted_at = 0 AND notebook_id IS NULL"
            ListFilter.Favorite -> where += "deleted_at = 0 AND is_favorite = 1"
            ListFilter.Deleted -> where += "deleted_at != 0"
            is ListFilter.Folder -> {
                where += "deleted_at = 0 AND notebook_id IN (SELECT id FROM notebooks WHERE folder_id = ? AND deleted_at = 0)"
                args += filter.folderId.toString()
            }
            is ListFilter.Notebook -> {
                where += "deleted_at = 0 AND notebook_id = ?"
                args += filter.notebookId.toString()
            }
        }
        val selection = where.joinToString(" AND ")
        val cursor = dbHelper.readableDatabase.rawQuery(
            "SELECT COUNT(*) FROM notes WHERE $selection",
            args.toTypedArray(),
        )
        cursor.use { c -> if (c.moveToFirst()) c.getInt(0) else 0 }
    }

    suspend fun setFavorite(id: Long, favorite: Boolean) = withContext(Dispatchers.IO) {
        val cv = ContentValues().apply {
            put("is_favorite", if (favorite) 1 else 0)
            put("updated_at", System.currentTimeMillis())
        }
        dbHelper.writableDatabase.update("notes", cv, "id = ?", arrayOf(id.toString()))
        Unit
    }

    suspend fun moveNoteToNotebook(noteId: Long, targetNotebookId: Long?) = withContext(Dispatchers.IO) {
        val cv = ContentValues()
        if (targetNotebookId == null) cv.putNull("notebook_id") else cv.put("notebook_id", targetNotebookId)
        cv.put("updated_at", System.currentTimeMillis())
        dbHelper.writableDatabase.update("notes", cv, "id = ?", arrayOf(noteId.toString()))
        Unit
    }

    enum class SortBy { UPDATED_DESC, CREATED_DESC }

    sealed class ListFilter {
        object All : ListFilter()
        object Uncategorized : ListFilter()
        object Favorite : ListFilter()
        object Deleted : ListFilter()
        data class Folder(val folderId: Long) : ListFilter()
        data class Notebook(val notebookId: Long) : ListFilter()
    }

    // ----- private -----

    private fun cursorToNote(c: Cursor): Note {
        val contentJson = c.getString(c.getColumnIndexOrThrow("content_json")) ?: ""
        val catIdx = c.getColumnIndexOrThrow("category_id")
        return Note(
            id = c.getLong(c.getColumnIndexOrThrow("id")),
            title = c.getString(c.getColumnIndexOrThrow("title")) ?: "",
            plainText = c.getString(c.getColumnIndexOrThrow("plain_text")) ?: "",
            isFavorite = c.getInt(c.getColumnIndexOrThrow("is_favorite")) == 1,
            createdAt = c.getLong(c.getColumnIndexOrThrow("created_at")),
            updatedAt = c.getLong(c.getColumnIndexOrThrow("updated_at")),
            content = NoteJson.fromJson(contentJson),
            categoryId = if (c.isNull(catIdx)) null else c.getLong(catIdx),
            deletedAt = c.getLong(c.getColumnIndexOrThrow("deleted_at")),
            notebookId = c.getColumnIndex("notebook_id").let { idx ->
                if (idx < 0 || c.isNull(idx)) null else c.getLong(idx)
            },
        )
    }
}

File: code/HuaweiNoteFlutter/lib/repositories/note_repository.dart
import 'package:sqflite/sqflite.dart';
import '../db/database_helper.dart';
import '../models/note.dart';
import '../models/note_content.dart';
import '../services/search_service.dart';

enum NoteSortBy { updatedDesc, createdDesc }

sealed class NoteListFilter {
  const NoteListFilter();
  static const all = AllFilter();
  static const uncategorized = UncategorizedFilter();
  static const favorite = FavoriteFilter();
  static const deleted = DeletedFilter();
  static NoteListFilter folder(int folderId) => FolderFilter(folderId);
  static NoteListFilter notebook(int notebookId) => NotebookFilter(notebookId);
}

class AllFilter extends NoteListFilter {
  const AllFilter();
}

class UncategorizedFilter extends NoteListFilter {
  const UncategorizedFilter();
}

class FavoriteFilter extends NoteListFilter {
  const FavoriteFilter();
}

class DeletedFilter extends NoteListFilter {
  const DeletedFilter();
}

class FolderFilter extends NoteListFilter {
  final int folderId;
  const FolderFilter(this.folderId);
}

class NotebookFilter extends NoteListFilter {
  final int notebookId;
  const NotebookFilter(this.notebookId);
}

class NoteRepository {
  final DatabaseHelper _dbHelper;

  NoteRepository(this._dbHelper);

  Future<List<Note>> list({
    NoteListFilter filter = const AllFilter(),
    NoteSortBy sortBy = NoteSortBy.updatedDesc,
    String? query,
  }) async {
    final db = await _dbHelper.database;
    final orderBy = switch (sortBy) {
      NoteSortBy.updatedDesc => 'updated_at DESC',
      NoteSortBy.createdDesc => 'created_at DESC',
    };
    final where = <String>[];
    final args = <String>[];
    switch (filter) {
      case AllFilter():
        where.add('deleted_at = 0');
      case UncategorizedFilter():
        where.add('deleted_at = 0 AND notebook_id IS NULL');
      case FavoriteFilter():
        where.add('deleted_at = 0 AND is_favorite = 1');
      case DeletedFilter():
        where.add('deleted_at != 0');
      case FolderFilter(:final folderId):
        where.add(
            'deleted_at = 0 AND notebook_id IN (SELECT id FROM notebooks WHERE folder_id = ? AND deleted_at = 0)');
        args.add(folderId.toString());
      case NotebookFilter(:final notebookId):
        where.add('deleted_at = 0 AND notebook_id = ?');
        args.add(notebookId.toString());
    }
    if (query != null && query.isNotEmpty) {
      final like = '%$query%';
      where.add('(title LIKE ? OR plain_text LIKE ?)');
      args.addAll([like, like]);
    }
    final rows = await db.query(
      'notes',
      where: where.join(' AND '),
      whereArgs: args,
      orderBy: orderBy,
    );
    return rows.map(_rowToNote).toList();
  }

  Future<Note?> get(int id) async {
    final db = await _dbHelper.database;
    final rows = await db.query('notes', where: 'id = ?', whereArgs: [id]);
    if (rows.isEmpty) return null;
    return _rowToNote(rows.first);
  }

  Future<List<Note>> getByIds(List<int> ids) async {
    if (ids.isEmpty) return [];
    final db = await _dbHelper.database;
    final placeholders = List.filled(ids.length, '?').join(',');
    final rows = await db.query(
      'notes',
      where: 'id IN ($placeholders) AND deleted_at = 0',
      whereArgs: ids,
    );
    return rows.map(_rowToNote).toList();
  }

  Future<List<Note>> searchByLike(String keyword, {int limit = 50}) async {
    if (keyword.isEmpty) return [];
    final db = await _dbHelper.database;
    final pattern = '%$keyword%';
    final rows = await db.query(
      'notes',
      where: 'deleted_at = 0 AND (title LIKE ? OR plain_text LIKE ?)',
      whereArgs: [pattern, pattern],
      orderBy: 'updated_at DESC',
      limit: limit,
    );
    return rows.map(_rowToNote).toList();
  }

  Future<int> save(Note note) async {
    final db = await _dbHelper.database;
    final now = DateTime.now().millisecondsSinceEpoch;

    final firstImage = _extractFirstImagePath(note.content);
    final hasTodo = _hasTodoBlock(note.content);

    final values = <String, Object?>{
      'title': note.title,
      'plain_text': note.content.toPlainText(),
      'content_json': note.content.toJson(),
      'is_favorite': note.isFavorite ? 1 : 0,
      'updated_at': now,
      'category_id': note.categoryId,
      'deleted_at': note.deletedAt,
      'notebook_id': note.notebookId,
      'background': note.background,
      'first_image_path': firstImage,
      'has_todo': hasTodo ? 1 : 0,
    };
    int resultId;
    if (note.id == 0) {
      values['created_at'] = note.createdAt > 0 ? note.createdAt : now;
      resultId = await db.insert('notes', values);
    } else {
      await db.update('notes', values, where: 'id = ?', whereArgs: [note.id]);
      resultId = note.id;
    }
    if (note.deletedAt == 0) {
      SearchService.instance.indexNote(note.copyWith(id: resultId));
    } else {
      SearchService.instance.removeNote(resultId);
    }
    return resultId;
  }

  Future<void> softDelete(int id) async {
    final db = await _dbHelper.database;
    await db.update(
      'notes',
      {'deleted_at': DateTime.now().millisecondsSinceEpoch},
      where: 'id = ?',
      whereArgs: [id],
    );
    SearchService.instance.removeNote(id);
  }

  Future<void> softDeleteBatch(List<int> ids) async {
    final db = await _dbHelper.database;
    final now = DateTime.now().millisecondsSinceEpoch;
    await db.transaction((txn) async {
      for (final id in ids) {
        await txn.update('notes', {'deleted_at': now},
            where: 'id = ?', whereArgs: [id]);
      }
    });
  }

  Future<void> restore(int id) async {
    final db = await _dbHelper.database;
    await db.update('notes', {'deleted_at': 0},
        where: 'id = ?', whereArgs: [id]);
    final note = await get(id);
    if (note != null) SearchService.instance.indexNote(note);
  }

  Future<void> deletePermanently(int id) async {
    final db = await _dbHelper.database;
    await db.delete('notes', where: 'id = ?', whereArgs: [id]);
    SearchService.instance.removeNote(id);
  }

  Future<int> count({NoteListFilter filter = const AllFilter()}) async {
    final db = await _dbHelper.database;
    final where = <String>[];
    final args = <String>[];
    switch (filter) {
      case AllFilter():
        where.add('deleted_at = 0');
      case UncategorizedFilter():
        where.add('deleted_at = 0 AND notebook_id IS NULL');
      case FavoriteFilter():
        where.add('deleted_at = 0 AND is_favorite = 1');
      case DeletedFilter():
        where.add('deleted_at != 0');
      case FolderFilter(:final folderId):
        where.add(
            'deleted_at = 0 AND notebook_id IN (SELECT id FROM notebooks WHERE folder_id = ? AND deleted_at = 0)');
        args.add(folderId.toString());
      case NotebookFilter(:final notebookId):
        where.add('deleted_at = 0 AND notebook_id = ?');
        args.add(notebookId.toString());
    }
    final result = await db.rawQuery(
      'SELECT COUNT(*) FROM notes WHERE ${where.join(' AND ')}',
      args,
    );
    return Sqflite.firstIntValue(result) ?? 0;
  }

  Future<void> setFavorite(int id, bool favorite) async {
    final db = await _dbHelper.database;
    await db.update(
        'notes',
        {
          'is_favorite': favorite ? 1 : 0,
          'updated_at': DateTime.now().millisecondsSinceEpoch,
        },
        where: 'id = ?',
        whereArgs: [id]);
  }

  Future<void> moveNoteToNotebook(int noteId, int? targetNotebookId) async {
    final db = await _dbHelper.database;
    await db.update(
        'notes',
        {
          'notebook_id': targetNotebookId,
          'updated_at': DateTime.now().millisecondsSinceEpoch,
        },
        where: 'id = ?',
        whereArgs: [noteId]);
  }

  Future<int> purgeExpired({
    int? now,
    int ttlMs = 30 * 24 * 60 * 60 * 1000,
  }) async {
    final db = await _dbHelper.database;
    final cutoff = (now ?? DateTime.now().millisecondsSinceEpoch) - ttlMs;
    return db.delete('notes',
        where: 'deleted_at > 0 AND deleted_at < ?', whereArgs: [cutoff]);
  }

  static String? _extractFirstImagePath(NoteContent content) {
    final doc = content.documentJson['document'] as Map<String, dynamic>?;
    if (doc == null) return null;
    final children = doc['children'] as List<dynamic>? ?? [];
    for (final node in children) {
      final map = node as Map<String, dynamic>;
      if (map['type'] == 'image') {
        final data = map['data'] as Map<String, dynamic>?;
        if (data != null) {
          return data['url'] as String?;
        }
      }
    }
    return null;
  }

  static bool _hasTodoBlock(NoteContent content) {
    final doc = content.documentJson['document'] as Map<String, dynamic>?;
    if (doc == null) return false;
    final children = doc['children'] as List<dynamic>? ?? [];
    for (final node in children) {
      final map = node as Map<String, dynamic>;
      if (map['type'] == 'todo_list') return true;
    }
    return false;
  }

  Note _rowToNote(Map<String, Object?> row) {
    final contentJson = row['content_json'] as String? ?? '';
    return Note(
      id: row['id'] as int,
      title: row['title'] as String? ?? '',
      plainText: row['plain_text'] as String? ?? '',
      isFavorite: (row['is_favorite'] as int) == 1,
      createdAt: row['created_at'] as int,
      updatedAt: row['updated_at'] as int,
      content: NoteContent.fromJson(contentJson),
      categoryId: row['category_id'] as int?,
      deletedAt: row['deleted_at'] as int? ?? 0,
      notebookId: row['notebook_id'] as int?,
      background: row['background'] as String? ?? 'plain',
      firstImagePath: row['first_image_path'] as String?,
      hasTodo: (row['has_todo'] as int? ?? 0) == 1,
    );
  }
}
